### ISSUE TITLE
Stack overflow when using type_alias_impl_trait with async functions

### ISSUE THREAD

**Issue by alexkirsz** [rust-lang/rust-analyzer#17403]
**rust-analyzer version**: rust-analyzer 0.0.0 (72dfbe95d 2024-06-10)

**rustc version**: rustc 1.80.0-nightly (1871252fc 2024-05-15)

**editor or extension**: Zed

**code snippet to reproduce**: 
```rust
#![feature(type_alias_impl_trait)]

type TypeAlias = impl Clone;

async fn alias() -> TypeAlias {
    todo!()
}
```

This causes an infinite recursion in `insert_inference_vars_for_impl_trait`, eventually resulting in a stack overflow which crashes RA.

My very naive fix for this issue was to add a set of already seen types to avoid `fold_tys` recursing, but this causes other issues down the road:

```
stderr: 2024-06-12T08:06:33.796309Z ERROR hir_ty::infer: Missed RPIT in `insert_inference_vars_for_rpit`
stderr: Panic context:
stderr: > fetch_native_diagnostics
stderr:
stderr: thread 'Worker' panicked at crates/hir/src/semantics.rs:1516:13:
stderr:
stderr:
stderr: Failed to lookup BLOCK_EXPR@96..111 in this Semantics.
stderr: Make sure to use only query nodes, derived from this instance of Semantics.
stderr: root node:   SOURCE_FILE@0..112
stderr: known nodes: SOURCE_FILE@0..112
stderr:
stderr:
stderr: stack backtrace:
stderr: 0: _rust_begin_unwind
stderr: 1: core::panicking::panic_fmt
stderr: 2: hir::semantics::SemanticsImpl::find_file
stderr: 3: hir::semantics::SemanticsImpl::analyze_impl
stderr: 4: hir::semantics::SemanticsImpl::scope
stderr: 5: ide_diagnostics::handlers::type_mismatch::type_mismatch
stderr: 6: ide_diagnostics::diagnostics
stderr: 7: salsa::Cancelled::catch
stderr: 8: ide::Analysis::diagnostics
stderr: 9: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
stderr: 10: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
stderr: 11: rust_analyzer::diagnostics::fetch_native_diagnostics
stderr: 12: core::ops::function::FnOnce::call_once{{vtable.shim}}
stderr: note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by alexkirsz** [comment]
Bisected and this started crashing with https://github.com/rust-lang/rust-analyzer/pull/16852

**Comment by alexkirsz** [comment]
Here's my naive workaround if anyone else is blocked on this issue:
https://github.com/rust-lang/rust-analyzer/commit/7299e6af09d4f6ecfaa23e6384b1e73c1175da07

**Comment by kmicklas** [comment]
> Here's my naive workaround if anyone else is blocked on this issue: [7299e6a](https://github.com/rust-lang/rust-analyzer/commit/7299e6af09d4f6ecfaa23e6384b1e73c1175da07)

Thank you! I'm hitting a variation of this issue (no async) and this has restored my productivity.

**Comment by kmicklas** [comment]
Looks like this is a duplicate of https://github.com/rust-lang/rust-analyzer/issues/17199.

**Comment by ShoyuVanilla** [comment]
😱 @rustbot claim

**Comment by ShoyuVanilla** [comment]
I found the root cause and solution. I'll make a PR when I get back home in few hours.
Sorry for the inconvenience to those who suffered from this issue 😢

**Comment by alexkirsz** [comment]
All good @ShoyuVanilla, thank you for your work on RA!
